File: go/min_path_sum.go
package main

import (
	"fmt"
	"math"
)

func min(a int, b int) int {
	if a < b {
		return a
	}
	return b
}

func minPathSum(grid [][]int) int {
	l := len(grid)
	c := len(grid[0])
	copy := make([][]int, l+1)
	for i := 0; i <= l; i++ {
		copy[i] = make([]int, c+1)
	}

	for i := 0; i < c+1; i++ {
		copy[0][i] = math.MaxInt64
	}
	for i := 0; i < l+1; i++ {
		copy[i][0] = math.MaxInt64
	}
	copy[1][1] = grid[0][0]
	for i := 1; i < l+1; i++ {
		for j := 1; j < c+1; j++ {
			if i == 1 && j == 1 {
				continue
			}
			copy[i][j] = grid[i-1][j-1] + min(copy[i-1][j], copy[i][j-1])
		}
	}

	return copy[l][c]
}

func main() {
	triangle := make([][]int, 3)
	for i := 0; i < len(triangle); i++ {
		triangle[i] = make([]int, 3)
	}
	triangle[0][0] = 1
	triangle[0][1] = 3
	triangle[0][2] = 1
	triangle[1][0] = 1
	triangle[1][1] = 5
	triangle[1][2] = 1
	triangle[2][0] = 4
	triangle[2][1] = 2
	triangle[2][2] = 1

	ret := minPathSum(triangle)
	fmt.Println(ret)
}

File: cc/dp/min_path_sum.cc
#include <iostream>
#include <vector>
#include <climits>
#include <algorithm>
#include <functional>

using namespace std;

#if  0
func minPathSum(grid [][]int) int {
	l := len(grid)
	if l == 0 {
		return 0
	}
	copy := make([][]int, l)
	c := len(grid[0])
	for i := 0; i < len(copy); i++ {
		copy[i] = make([]int, c)
		//	note[i] = make([]int, i+1)
	}
	for _, v := range copy{
		for i := 0; i < len(v); i++{
			v[i] = -1
		}
	}

	var helper func(int, int) int
	helper = func(i int, j int) int{
/*  	if i == 0 || j == 0 {
		if i == 0 && j > 0 {
			return grid[i][j] + helper(i, j-1)
		} else if i > 0 && j == 0 {
			return grid[i][j] + helper(i-1, j)
		} else {
			return grid[i][j]
		}
	}  */
	if i < 0 || j < 0{
		return math.MaxInt64
	}
	if i == 0 && j == 0{
		return grid[i][j]
	}

	if copy[i][j] == -1{
		copy[i][j] = grid[i][j] + min(helper(i, j-1), help
                       
#endif 
class Solution {
public:
    int minPathSum(vector<vector<int>>& grid) {
           auto l  = grid.size();
           auto c = grid[0].size();
           vector<vector<int>> copy(l, vector<int>(c, -1));

            std::function<int(int, int)>  helper;
            helper  =  [&] (int  i, int j)->int{
               if  (i  <  0  ||  j  <  0){
                   return  INT_MAX; 
               } 
               if  (i==0 && j == 0){
                    return grid[i][j];
               }
              
              if( copy[i][j] == -1){
		            copy[i][j] = grid[i][j] + std::min(helper(i, j-1), helper(i-1, j));
              }
              return copy[i][j] ;
            };
            
            return helper(l-1, c-1);
    }
};

int main(){
    return 0;
}